src/kernel/disconnected_transactions.h

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2023 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_KERNEL_DISCONNECTED_TRANSACTIONS_H
+#define BITCOIN_KERNEL_DISCONNECTED_TRANSACTIONS_H
+
+#include <core_memusage.h>
+#include <memusage.h>
+#include <primitives/transaction.h>
+#include <util/hasher.h>
+
+#include <list>
+#include <unordered_map>
+
+/** Maximum kilobytes for transactions to store for processing during reorg */
+static const unsigned int MAX_DISCONNECTED_TX_POOL_SIZE = 20'000;
+/**
+ * DisconnectedBlockTransactions
+
+ * During the reorg, it's desirable to re-add previously confirmed transactions
+ * to the mempool, so that anything not re-confirmed in the new chain is
+ * available to be mined. However, it's more efficient to wait until the reorg
+ * is complete and process all still-unconfirmed transactions at that time,
+ * since we expect most confirmed transactions to (typically) still be
+ * confirmed in the new chain, and re-accepting to the memory pool is expensive
+ * (and therefore better to not do in the middle of reorg-processing).
+ * Instead, store the disconnected transactions (in order!) as we go, remove any
+ * that are included in blocks in the new chain, and then process the remaining
+ * still-unconfirmed transactions at the end.
+ *
+ * Order of queuedTx:
+ * The front of the list should be the most recently-confirmed transactions (transactions at the
+ * end of vtx of blocks closer to the tip). If memory usage grows too large, we trim from the front
+ * of the list. After trimming, transactions can be re-added to the mempool from the back of the
+ * list to the front without running into missing inputs.
+ */
+class DisconnectedBlockTransactions {
+private:
+    uint64_t cachedInnerUsage = 0;
+    const size_t m_max_mem_usage;
+    std::list<CTransactionRef> queuedTx;
+    using TxList = decltype(queuedTx);
+    std::unordered_map<uint256, TxList::iterator, SaltedTxidHasher> iters_by_txid;
+
+    /** Trim the earliest-added entries until we are within memory bounds. */
+    std::vector<CTransactionRef> LimitMemoryUsage()
+    {
+        std::vector<CTransactionRef> evicted;
+
+        while (!queuedTx.empty() && DynamicMemoryUsage() > m_max_mem_usage) {
+            evicted.emplace_back(queuedTx.front());
+            cachedInnerUsage -= RecursiveDynamicUsage(queuedTx.front());
```

Comment:
I think `cachedInnerUsage` shouldn't account for the memory usage of the `shared_ptr`, only for the `CTransaction`. We already account for the `shared_ptr` in `DisconnectedBlockTransactions::DynamicMemoryUsage` (as I think it should be).

<details>
<summary>git diff</summary>

```diff
diff --git a/src/kernel/disconnected_transactions.h b/src/kernel/disconnected_transactions.h
index 0c170e7a4b..a3bc77b6fc 100644
--- a/src/kernel/disconnected_transactions.h
+++ b/src/kernel/disconnected_transactions.h
@@ -50,7 +50,7 @@ private:
 
         while (!queuedTx.empty() && DynamicMemoryUsage() > m_max_mem_usage) {
             evicted.emplace_back(queuedTx.front());
-            cachedInnerUsage -= RecursiveDynamicUsage(queuedTx.front());
+            cachedInnerUsage -= RecursiveDynamicUsage(*queuedTx.front());
             iters_by_txid.erase(queuedTx.front()->GetHash());
             queuedTx.pop_front();
         }
@@ -91,7 +91,7 @@ public:
         for (auto block_it = vtx.rbegin(); block_it != vtx.rend(); ++block_it) {
             auto it = queuedTx.insert(queuedTx.end(), *block_it);
             iters_by_txid.emplace((*block_it)->GetHash(), it);
-            cachedInnerUsage += RecursiveDynamicUsage(*block_it);
+            cachedInnerUsage += RecursiveDynamicUsage(**block_it);
         }
         return LimitMemoryUsage();
     }
@@ -108,7 +108,7 @@ public:
             if (iter != iters_by_txid.end()) {
                 auto list_iter = iter->second;
                 iters_by_txid.erase(iter);
-                cachedInnerUsage -= RecursiveDynamicUsage(*list_iter);
+                cachedInnerUsage -= RecursiveDynamicUsage(**list_iter);
                 queuedTx.erase(list_iter);
             }
         }

```
</details>

Reply:
I agree, changed (but in the rewrite commit instead of here)

Reply:
Revisiting this again, I think my suggestion was wrong and we're now underestimating memory usage.

In `DisconnectedBlockTransactions::DynamicMemoryUsage`, with `memusage::DynamicUsage(queuedTx)` we account only for the _dynamic_ allocation of `queuedTx`, which is `(2 ptrs + sizeof(CTransactionRef)) * size`. This corresponds to 2+2 pointers per item, and does _not_ include the managed `CTransaction` object or the control block.

In `AddTransactionsFromBlock` (and other places where we modify `cachedInnerUsage`), when we calculate `RecursiveDynamicUsage(**block_it)` (i.e. on a `CTransaction`) as I suggested here, we only account for the `vins` and `vouts` of the transaction, and not for the `CTransaction` object nor the control block.

So effectively, we do need to account for the `CTransactionRef` in both places, because in `DynamicMemoryUsage` we account for the 2 pointers, and in `AddTransactionRef` we account for the managed object and the control block. In the current implementation, we are not accounting for the `CTransaction` or the control block, even though everything is allocated on the heap.

---

To illustrate, for a coinbase transaction, the difference in `AddTransactionsFromBlock` is:
```
(lldb) p (size_t) RecursiveDynamicUsage(*block_it)
(size_t) 640
(lldb) p (size_t) RecursiveDynamicUsage(**block_it)
(size_t) 448
```

cc @ismaelsadeeq perhaps something to include in #28530?

Reply:
Fixed in https://github.com/stickies-v/bitcoin/commit/87f3f28b3acfc329eaed0b0cce6ff589271bfefd - happy to open a separate pull if you prefer though.

Reply:
Thanks @stickies-v  happy to add it to #28530, I f I understand you correctly, you mean that we are not accounting for the `shared_ptr` and the `CTransaction` object before only the `vins` and `vouts`.

>In DisconnectedBlockTransactions::DynamicMemoryUsage, with memusage::DynamicUsage(queuedTx) we account only for the dynamic allocation of queuedTx, which is (2 ptrs + sizeof(CTransactionRef)) * size. This corresponds to 2+2 pointers per item.


I think this is correct.
We only account for the `vins` and `vouts`, hence the `CTransaction` Object itself and the `shared_ptr` that makes up `CTransactionRef` are not accounted.

With https://github.com/stickies-v/bitcoin/commit/87f3f28b3acfc329eaed0b0cce6ff589271bfefd we are now accounting for the `shared_ptr` `CTransaction` Object, its inputs and outputs and the dynamic allocation of `queuedTx` in .

because we pass shared_ptr<CTransaction> which executes  the below version of RecursiveDynamicUsage
```cpp
template<typename X>
static inline size_t RecursiveDynamicUsage(const std::shared_ptr<X>& p) {
    return p ? memusage::DynamicUsage(p) + RecursiveDynamicUsage(*p) : 0;
}
```
`memusage::DynamicUsage(p)` accounts for the size of `p` which is `CTransaction` and then accounts for the shared pointer size.

It is unclear to me what
>does not include the managed CTransaction object or the control block.
mean

Do you mean the `shared_ptr` and `CTransaction` Object ?

Reply:
> `RecursiveDynamicUsage(*p)` accounts for the vins and vouts,

We don't want to call `RecursiveDynamicUsage(*p)` anymore, only `RecursiveDynamicUsage(p)` as per my suggested fix

> whereas the two raw pointers are being accounted in `cachedInnerUsage`.

Not quite: the raw pointers are accounted for in `DisconnectedBlockTransactions::DynamicMemoryUsage` when calling `memusage::DynamicUsage(queuedTx)`.

Reply:
Yeah
But calling `RecursiveDynamicUsage(p)` means calling
```cpp
template<typename X>
static inline size_t RecursiveDynamicUsage(const std::shared_ptr<X>& p) {
    return p ? memusage::DynamicUsage(p) + RecursiveDynamicUsage(*p) : 0;
}
```
which brings to my point that 

    - `memusage::DynamicUsage(p)` accounts for the CTransaction Class and the control block.
    - `RecursiveDynamicUsage(*p)` accounts for the `vin`s and `vout`s,
    
> Not quite: the raw pointers are accounted for in DisconnectedBlockTransactions::DynamicMemoryUsage when calling memusage::DynamicUsage(queuedTx).
Yes.